src/cv/aggregator.cpp

```diff
@@ -0,0 +1,95 @@
+#include "cv/aggregator.hpp"
+#include "utilities/locks.hpp"
+#include "utilities/constants.hpp"
+#include "utilities/logging.hpp"
+#include "utilities/lockptr.hpp"
+
+CVAggregator::CVAggregator(Pipeline p): pipeline{p} {
+    this->num_worker_threads = 0;
+    this->results = std::make_shared<CVResults>();
+    for (int i = 0; i < this->results->matches.size(); i++) {
+        this->results->matches[i] = -1;
+    }
+}
+
+CVAggregator::~CVAggregator() {
+
+}
+
+LockPtr<CVResults> CVAggregator::getResults() {
+    return LockPtr<CVResults>(this->results, &this->mut);
+}
+
+void CVAggregator::runPipeline(const ImageData& image) {
+    Lock lock(this->mut);
+
+    if (this->num_worker_threads > MAX_CV_PIPELINES) {
+        LOG_F(WARNING, "Tried to spawn more than %ld CVAggregator workers at one time.", MAX_CV_PIPELINES);
+        this->overflow_queue.push(image);
+        LOG_F(WARNING, "CVAggregator overflow queue at size %ld", this->overflow_queue.size());
+        return;
+    }
+
+    static int thread_num = 0;
+    this->num_worker_threads++;
+    std::thread worker(&CVAggregator::worker, this, image, ++thread_num);
+    worker.detach();
+}
+
+void CVAggregator::worker(ImageData image, int thread_num) {
+    loguru::set_thread_name(("cv worker " + std::to_string(thread_num)).c_str());
+    LOG_F(INFO, "New CVAggregator worker spawned.");
+
+    while (true) {
+        auto results = this->pipeline.run(image);
+        
+        Lock lock(this->mut);
```

Comment:
I do not understand the behavior of the CVAggregator mut. What is this protecting? I also am unsure how this mut behaves when you have threads.

Reply:
Each worker thread is able to run the pipeline independently because they are just accessing the models and running the algorithms on the images independently, but once the results have been extracted from the pipeline the worker thread needs to put the results back in the two state variables in CVAggregator: the vector of all detected targets and the array containing match information. So this logic of going into the CVAggregator data members and modifying the state needs to be a critical section so that two workers don't try and modify the aggregated results at the same time.

Reply:
Each worker thread is running the private function "worker", so they all have access to the shared memory contained within the `this` pointer. So they all can lock the same mutex and append to the same state variables inside of CVAggregator.